nep-12.mediawiki

```diff
@@ -0,0 +1,2130 @@
+      break;
+    case INSUFFICIENT_FUNDS:
+      console.log('The user has insufficient funds to execute this transaction.');
+      break;
+  }
+});
+</pre>
+
+
+'''Provider Request Handling'''
+
+Upon receiving this request
+* Validate the inputs
+** From address matches connected account
+** To address is valid
+** Validate asset exists
+** Amount is valid value, and account has enough balance
+* If the "broadcastOverride" input argument is set to True, return the serialized signed transaction to the dApp
+* Else broadcast the RPC request
+* Return the transaction Id and the url of the rpc node which the tx was broadcast to
+
+Example
+<pre>
+{
+  txid: 'ed54fb38dff371be6e3f96e4880405758c07fe6dd1295eb136fe15f311e9ff77',
+  nodeUrl: 'http://seed7.ngd.network:10332',
+}
+</pre>
+
+
+====invoke====
+Execute a contract invocation.
+
+
+'''Method Interface'''
+
+<pre>
+function invoke({
+  scriptHash: string,
+  operation: string,
+  args: Argument[],
+  attachedAssets?: AttachedAssets,
+  fee?: string,
+  assetIntentOverrides?: AssetIntentOverrides,
+  triggerContractVerification?: boolean,
+  network?: string,
+  broadcastOverride?: boolean,
+  txHashAttributes?: TxHashAttribute[],
+}): Promise<InvokeOutput>
+</pre>
+
+
+'''Input arguments'''
+
+<pre>
+interface InvokeArgs {
+  scriptHash: string; // script hash of the smart contract to invoke
+  operation: string; // operation on the smart contract to call
+  args: Argument[]; // any input arguments for the operation
+  fee?: string; // (Optional) The parsed amount of network fee (in GAS) to include with transaction
+  network?: string - Network to submit this request to. If omitted, will default to network the wallet is currently set to.
+  attachedAssets?: AttachedAssets;
+  broadcastOverride?: boolean; // In the case that the dApp would like to be responsible for broadcasting the signed transaction rather than the wallet provider
+
+  assetIntentOverrides?: AssetIntentOverrides;
+  // A hard override of all transaction utxo inputs and outputs.
+  // IMPORTANT: If provided, fee and attachedAssets will be ignored.
+
+  triggerContractVerification?: boolean; // Adds the instruction to invoke the contract verification trigger
+
+  txHashAttributes?: TxHashAttribute[]; // Adds transaction attributes for the "Hash<x>" usage block
+}
+
+interface Argument {
+  type: ArgumentDataType;
+  value: any;
+}
+
+interface TxHashAttribute extends Argument {
+  txAttrUsage: 'Hash1'|'Hash2'|'Hash3'|'Hash4'|'Hash5'|'Hash6'|'Hash7'|'Hash8'|'Hash9'|'Hash10'|'Hash11'|'Hash12'|'Hash13'|'Hash14'|'Hash15';
+}
+
+type ArgumentDataType = 'String'|'Boolean'|'Hash160'|'Hash256'|'Integer'|'ByteArray'|'Array'|'Address';
+
+interface AttachedAssets {
+  [asset: 'NEO'|'GAS']: string;
+}
+// KEY: Asset symbol (only NEO or GAS)
+// VALUE: Parsed amount to attach
+
+interface AssetIntentOverrides {
+  inputs: AssetInput[];
+  outputs: AssetOutput[];
+}
+
+interface AssetInput {
```

Comment:
could we name this more precisely, like, UTXOAssetInput? So we let it clear it's about utxo.

Reply:
updated.